### PR TITLE
Feature/Attribute deletion

### PR DESCRIPTION
**Pull request by Milena-Czierlinski**

### CODE REVIEW

**Comment by jkoenig134** [comment]
@Mergifyio update

**Comment by Milena-Czierlinski** [comment]
@Mergifyio update

File: packages/consumption/src/modules/attributes/local/LocalAttribute.ts [hunk 1]
```diff
@@ -132,28 +143,32 @@ export class LocalAttribute extends CoreSynchronizable implements ILocalAttribut
...
 
+    public isThirdPartyOwnedAttribute(ownAddress: CoreAddress, thirdPartyAddress?: CoreAddress): this is ThirdPartyOwnedRelationshipAttribute {
+        let isThirdPartyOwnedAttribute = this.isShared() && !this.isOwnedBy(ownAddress) && !this.isOwnedBy(this.shareInfo.peer);
+
+        if (typeof thirdPartyAddress !== "undefined") {
+            isThirdPartyOwnedAttribute &&= this.isOwnedBy(thirdPartyAddress);
+        }
+        return isThirdPartyOwnedAttribute;
+    }
+
```
Comment (slavistan) [inline packages/consumption/src/modules/attributes/local/LocalAttribute.ts:163]: Third-party owned attributes can only be relationship attributes, right?

Comment (Milena-Czierlinski) [reply packages/consumption/src/modules/attributes/local/LocalAttribute.ts:163]: Yes, this is why we only define the type `ThirdPartyOwnedRelationshipAttribute` and have a check `isThirdPartyOwnedRelationshipAttribute`, which internally calls `isRelationshipAttribute() && isThirdPartyOwnedAttribute()`.

**Review by jkoenig134 — APPROVED** [review]
LGTM (as I already reviewed the PRs before I just skimmed it), I'd highly vote for waiting for more reviews especially @stnmtz and maybe @tnotheis also wants to give feedback :)

File: packages/consumption/test/modules/attributes/AttributesController.test.ts [hunk 2]
```diff
@@ -543,6 +532,62 @@ describe("AttributesController", function () {
...
+
+            await consumptionController.attributes.deleteAttribute(complexAttribute);
+
+            const deletedAttribute = await consumptionController.attributes.getLocalAttribute(complexAttribute.id);
+            expect(deletedAttribute).toBeUndefined();
+
+            for (const childAttribute of childAttributes) {
+                const deletedChildAttribute = await consumptionController.attributes.getLocalAttribute(childAttribute.id);
+                expect(deletedChildAttribute).toBeUndefined();
+            }
```
Comment (stnmtz) [inline packages/consumption/test/modules/attributes/AttributesController.test.ts:587]: Check event bus for attribute deleted event

**Review by stnmtz — CHANGES_REQUESTED** [review]

File: packages/consumption/test/modules/notifications/itemProcessors/attributeDeleted/PeerSharedAttributeDeletedByPeerNotificationItemProcessor.test.ts [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        expect(updatedAttribute!.deletionInfo?.deletionStatus).toStrictEqual(DeletionStatus.DeletedByPeer);
+
+        /* Manually trigger and verify rollback. */
+        await processor.rollback(notificationItem, notification);
+        const attributeAfterRollback = await consumptionController.attributes.getLocalAttribute(notificationItem.attributeId);
+        expect(attributeAfterRollback?.deletionInfo).toBeUndefined();
+    });
+
+    test("runs all processor methods for a succeeded attribute", async function () {
+        const predecessorOSRA = await consumptionController.attributes.createAttributeUnsafe({
```
[the remark this answers is not in the document]
Comment (sebbi08) [reply packages/consumption/test/modules/notifications/itemProcessors/attributeDeleted/PeerSharedAttributeDeletedByPeerNotificationItemProcessor.test.ts:168]: OwnSharedRelationshipAttribute ?

File: packages/runtime/src/events/EventProxy.ts [hunk 4]
```diff
@@ -76,6 +79,18 @@ export class EventProxy {
             this.targetEventBus.publish(new AttributeDeletedEvent(event.eventTargetAddress, AttributeMapper.toAttributeDTO(event.data)));
         });
 
+        this.subscribeToSourceEvent(consumption.OwnSharedAttributeDeletedByOwnerEvent, (event) => {
```
[the remark this answers is not in the document]
Comment (Milena-Czierlinski) [reply packages/runtime/src/events/EventProxy.ts:82]: This is also not unambiguous: Did the peer notify about the deletion of an own shared Attribute (of them) or was the peer notified about the deletion of an own shared Attribute (of me). Same below.

File: packages/consumption/src/modules/requests/itemProcessors/deleteAttribute/DeleteAttributeRequestItemProcessor.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+        const deletionInfo = LocalAttributeDeletionInfo.from({
+            deletionStatus: DeletionStatus.ToBeDeletedByPeer,
+            deletionDate: responseItem.deletionDate
+        });
+
+        const predecessors = await this.consumptionController.attributes.getPredecessorsOfAttribute(attribute.id);
+
+        for (const attr of [attribute, ...predecessors]) {
+            if (attr.deletionInfo?.deletionStatus !== DeletionStatus.DeletedByPeer) {
```
[the remark this answers is not in the document]
Comment (Milena-Czierlinski) [reply packages/consumption/src/modules/requests/itemProcessors/deleteAttribute/DeleteAttributeRequestItemProcessor.ts:116]: There are two tests for this in the "applyIncomingResponseItem" block in DeleteAttributeRequestItemProcessor.test.ts.

**Pull request merged**